step/step.go

```diff
@@ -42,21 +67,221 @@ func New(logger log.Logger, inputParser stepconf.InputParser, commandFactory com
+	saver := cache.NewRestorer(step.envRepo, step.logger, step.commandFactory)
+
+	var cacheKey = []string{
+		fmt.Sprintf(dockerCacheKeyTemplate, imageName),
+		fmt.Sprintf("docker-%s-{{ .OS }}-{{ .Arch }}-{{ .Branch }}", imageName),
+		fmt.Sprintf("docker-%s-{{ .OS }}-{{ .Arch }}", imageName),
+	}
+
+	return saver.Restore(cache.RestoreCacheInput{
+		StepId:  "restore-cache",
+		Verbose: input.Verbose,
+		Keys:    cacheKey,
+	})
+}
+
+func (step DockerBuildPushStep) saveCache(input Input, imageName string) error {
+	step.logger.Infof("Saving cache...")
+	saver := cache.NewSaver(step.envRepo, step.logger, step.pathProvider, step.pathModifier, step.pathChecker)
+
+	return saver.Save(cache.SaveCacheInput{
+		StepId:      "save-cache",
+		Verbose:     input.Verbose,
+		Key:         fmt.Sprintf(dockerCacheKeyTemplate, imageName),
+		Paths:       []string{dockerCacheFolder},
+		IsKeyUnique: false,
+	})
+}
+
+func (step DockerBuildPushStep) dockerBuild(input Input, imageName string) error {
+	step.logger.Infof("Building docker image...")
+
+	if err := step.createCacheFolder(dockerCacheFolder); err != nil {
+		return fmt.Errorf("create cache folder: %w", err)
+	}
+	if err := step.createCacheFolder(dockerCacheFolderTemporary); err != nil {
+		return fmt.Errorf("create cache folder: %w", err)
+	}
+
+	buildkitContainer, err := step.initializeBuildkit(input)
+	if err != nil {
+		return fmt.Errorf("initialize buildkit: %w", err)
+	}
+	defer func() {
+		if err := step.destroyContainer(buildkitContainer); err != nil {
+			step.logger.Errorf("destroy buildx instance: %s", err)
+		}
+	}()
+
+	if err := step.build(input, imageName); err != nil {
+		return fmt.Errorf("build docker image: %w", err)
+	}
+	if err := step.moveCacheFolder(); err != nil {
+		return fmt.Errorf("move cache folder: %w", err)
+	}
+
+	return nil
+}
+
+func (step DockerBuildPushStep) destroyContainer(container string) error {
+	args := []string{
+		"buildx", "rm", "--force", container,
+	}
+	cmd := step.commandFactory.Create("docker", args, nil)
+	out, err := cmd.RunAndReturnTrimmedCombinedOutput()
+	if err != nil {
+		return fmt.Errorf("remove buildx instance %s: %w", out, err)
+	}
+	return nil
+}
+
+func (step DockerBuildPushStep) build(input Input, imageName string) error {
+	stdout := NewLoggerWriter(step.logger)
+	defer stdout.Flush()
+
+	args := []string{
+		"buildx",
+		"build",
+	}
+
+	if input.BuildArg != "" {
+		for _, arg := range strings.Split(input.BuildArg, "\n") {
+			args = append(args, "--build-arg", arg)
+		}
+	}
+
+	switch {
+	case input.UseBitriseCache:
+		args = append(args, fmt.Sprintf("--cache-from=type=local,src=%s", dockerCacheFolder))
+		args = append(args, fmt.Sprintf("--cache-to=type=local,dest=%s,mode=max,compression=zstd", dockerCacheFolderTemporary))
+	case input.CacheFrom != "":
+		for _, cacheFrom := range strings.Split(input.CacheFrom, "\n") {
+			args = append(args, fmt.Sprintf("--cache-from=%s", cacheFrom))
+		}
+		fallthrough
```

Comment:
This fallthrough could look weird, and we apologize for it, but thinking through all the alternative implementations funnily enough this seemed the less complex :D 

Logic: 
- When UseBitriseCache is true, we only want to set our fix cache from and cache-to params
- Otherwise we want to set cache-from and cache-to individually only and only if they are not empty (because adding an empty value to the array will break the cmdrunner)

Reply:
This sounds reasonable to me, but then it should be documented that the cache inputs and the Bitrise cache input are mutually exclusive

Reply:
It is documented in the step.yml which should show up on the UI as well right?